docs: Component documentation

nick-1666 opened the pull request:
Added a page under the "advanced" tab documenting components. Heavily based off of the random page.

Closes #202

nick-1666 commented:
also `sputnick#1666` for green role

Tandashi commented on packages/examples/src/scenes/components-switch.tsx:
```diff
@@ -0,0 +1,85 @@
+  private readonly container = createRef<Rect>();
+
+  public constructor(props?: SwitchProps) {
+    super({
```
Super minor detail: Is there a reason why you create a new object and spread the `props` instead of just passing the props directly?

```ts
super(props);
```

ksassnowski replied:
This and the the empty `<>` when adding only a single root element also apply to a few more places in this file.

nick-1666 replied:
> Super minor detail: Is there a reason why you create a new object and spread the `props` instead of just passing the props directly?
> 
> ```ts
> super(props);
> ```

because if you were to make a layout component and wanted to ensure it was a layout, you would use 
```tsx
super({
   layout
   ...props
})
```

ksassnowski replied:
I'm not sure I'm following. Where is the `layout` variable coming from in this example?

nick-1666 replied:
because if you were to make a layout component and wanted to ensure it was a layout, you would use :
```tsx
super({
   layout
   ...props
})
```
just makes it a bit more obvious not only props can go here

ksassnowski replied:
If you were making a layout component, you'd also be extending from `Layout` and not `Node`, however. So the props in this case would also extends `LayoutProps` and not `NodeProps`. At which point `super(props)` would still work as expected.

Tandashi replied:
Yeah no I get that but was confused because in that concrete example it is not needed. But if you want to add it just to provide a generic way to make it easier understandable for someone that isn't so much into typescript / coding I am all for it. But providing an example / paragraph describing this could be useful then as well.

Not sure however at what persona the documentation is targeted. Maybe @aarthificial could shed some light on this.

Tandashi commented on packages/docs/docs/advanced/custom-components.mdx:
````diff
@@ -0,0 +1,126 @@
+    @initial("#68ABDF")
+    @signal()
+    public declare readonly accent: ColorSignal<this>
+```
````
Yet another super minor detail: Personally I am missing a `}`. Or 
```ts
// ...
}
```
Since you move on to a different section in the next code snipped.

nick-1666 replied:
okay. will be in the next commit in the name of consistency if its like this everywhere else

nick-1666 commented on packages/examples/src/scenes/components-switch.tsx:
```diff
@@ -0,0 +1,85 @@
+    this.isOn = this.initialState();
+
+    this.add(
+      <>
```
because if you were to make a layout component and wanted to ensure it was a layout, you would use :
```tsx
super({
   layout
   ...props
})
```
just makes it a bit more obvious not only props can go here

nick-1666 replied:
this is true, it was 11 and it slipped past me lol. will be fixed

nick-1666 commented:
@Tandashi fixed in new commits. Review?

hhenrichsen commented on packages/docs/docs/advanced/custom-components.mdx:
```diff
@@ -0,0 +1,126 @@
+<AnimationPlayer name="components" banner />
+
+Components are reusable parts of code such as `Rect` and `Circle` which can be
+reused throughout scenes. To use a component in a scene, add it to the view and
```
nit: reusable/reused is repeated

nick-1666 replied:
idk what i was thinking 💀

hhenrichsen commented:
As per the contributing guide, [any commits addressing code reviews should be fixup commits](https://github.com/motion-canvas/motion-canvas/blob/main/CONTRIBUTING.md#addressing-review-feedback). This helps us to keep git history clean, and tell a consistent story.

hhenrichsen commented on packages/docs/docs/advanced/custom-components.mdx:
````diff
@@ -0,0 +1,126 @@
+    accent?: SignalValue<PossibleColor>
+}
+```
+Next, create a component class for your components. It must extend an existing
````
I would start this section with the requirements (i.e., "must extend Node or one of its subclasses"), and then work towards the recommendation to pick the most closely associated subclass. This sets up a good flow from hard rules to recommendations.

hhenrichsen commented on packages/docs/docs/advanced/custom-components.mdx:
````diff
@@ -0,0 +1,126 @@
+```
+
+To use the properties defined in the interface, your class _must_ contain a property with the
+same name. Motion Canvas provides type decorators to facilitate this. Here is an example of
````
I would reword the middle portion of this to read "Motion Canvas provides type decorators to facilitate this, like `@signal()` and `@initial()`. For more info on Signals, look here." with a link to the signals page. You may also want to explicitly state that `@initial()` is optional below.

hhenrichsen commented on packages/docs/docs/advanced/custom-components.mdx:
````diff
@@ -0,0 +1,126 @@
+```
+Now the properties will follow the schema defined in your interface and they will be
+passed to the superclass, in our case `Node`. The `props` parameter can also be useful
+outside of the super call for instance, setting styles on every rect in an array.
````
nit: This may read better as 
> The `props` parameter can also be useful outside of the super call to access your data elsewhere. For example, if you wanted to set a color set in the `props` on every `Rect` in an array.

"For instance" is normally a suffix.

hhenrichsen commented on packages/docs/docs/advanced/custom-components.mdx:
````diff
@@ -0,0 +1,126 @@
+    )
+}
+```
+Since this is still a regular typescript class, you can add methods. This is especially useful
````
nit: "Since this is still a regular typescript class, you can also add methods."

"Since this is still..." normally implies some base level of information, so "also" makes sense here.

hhenrichsen reviewed (commented):
Not going to approve or deny since most of what I'm leaving is grammatical over content, but here are some of my thoughts on how we can make this a little cleaner.

Tandashi commented on packages/docs/docs/advanced/custom-components.mdx:
```diff
@@ -0,0 +1,135 @@
+public constructor(props?: SwitchProps) {
+    ...
+    this.add(
+        <>
```
nit: empty element `<>` which isn't really needed.

nick-1666 replied:
Fixed this code sample in the new commit and changed the formatting to improve clarity

Tandashi commented:
Also a person opinion where others also are free to voice their opinion.

I **personally** like when things are linked. You used a lot of classes like `Layout`, `Node`, `Circle`, `Rect`, ...
I think it would be a good idea to make them clickable and link them to the respective API Reference.
If I am not mistaken that is how other documentation pages handle this as well.

nick-1666 commented:
> As per the contributing guide, [any commits addressing code reviews should be fixup commits](https://github.com/motion-canvas/motion-canvas/blob/main/CONTRIBUTING.md#addressing-review-feedback). This helps us to keep git history clean, and tell a consistent story.

fixed!

hhenrichsen commented:
It looks like something went really wrong in the process because there are now 13 commits on this PR instead of the 2 or 3 there should be.

Tandashi commented:
> > As per the contributing guide, [any commits addressing code reviews should be fixup commits](https://github.com/motion-canvas/motion-canvas/blob/main/CONTRIBUTING.md#addressing-review-feedback). This helps us to keep git history clean, and tell a consistent story.

> fixed!

Ehm... Shouldn't be 13 commits now. You probably merged the `main` into your branch I assume.

nick-1666 commented:
uh oh

hhenrichsen commented:
Also please go read the contributing guide. `fixup` commits are not the same as `fix:`-scoped commits. Mention me on the discord if you want me to walk you through how fixup commits work.

nick-1666 commented:
well uhh i didn't mean to spam request reviews. It was a visual glitch :/
my bad

hhenrichsen commented on packages/docs/docs/advanced/custom-components.mdx:
```diff
@@ -0,0 +1,141 @@
+
+<AnimationPlayer name="components" banner />
+
+Components are reusable parts of code such as `Rect` and `Circle` which can be
```
> Components are reusable parts of code such as `Rect` and `Circle` which can be scenes.

Looks like something got lost here 😅

nick-1666 replied:
Bro i swear i try not to be superficial but they just keep getting past me lol

hhenrichsen commented on packages/docs/docs/advanced/custom-components.mdx:
```diff
@@ -0,0 +1,141 @@
+
+<AnimationPlayer name="components" banner />
+
+Components are reusable parts of code such as `Rect` and `Circle` which can be
```
Additionally: did we want these to link to the API pages for `Rect` and `Circle`?

nick-1666 replied:
yes we do

hhenrichsen commented on packages/docs/docs/advanced/custom-components.mdx:
````diff
@@ -0,0 +1,141 @@
+    initialState={false}
+/>
+```
+To define what arguments a component will take in, first define an interface. All 
````
super nit: "in" is redundant here

nick-1666 replied:
youre right

hhenrichsen commented on packages/docs/docs/advanced/custom-components.mdx:
````diff
@@ -0,0 +1,141 @@
+    accent?: SignalValue<PossibleColor>
+}
+```
+Next, create a component class for your components. The component class must extend [Node](/api/2d/components/Node)
````
nit: "Next, create a ~~component~~ class for your components."

nick-1666 replied:
okay true

hhenrichsen commented on packages/docs/docs/advanced/custom-components.mdx:
````diff
@@ -0,0 +1,141 @@
+```
+Next, create a component class for your components. The component class must extend [Node](/api/2d/components/Node)
+or one of its subclasses. If you dont want to inherit any methods from an existing component, extend your class from
+`Node`. It is advisable to extend from a component most similar to the custom component you are building. For instance, 
````
We should probably keep with the plural first person here: "We advise extending from ~~a~~ the component most similar to the ~~custom~~ component you are building"

nick-1666 replied:
done!

hhenrichsen commented on packages/docs/docs/advanced/custom-components.mdx:
```diff
@@ -0,0 +1,141 @@
+export class Switch extends Node {
+    // @initial - sets the property to an
+    // initial value if it was not provided.
+    // The @initial decorator is optional.
```
Since this is on a comment on the `@initial` decorator, it's probably fine to say "This decorator is optional," or even update the beginning of this comment to "@initial - optional, sets the property to an..."

nick-1666 replied:
ill go with the latter

hhenrichsen commented on packages/docs/docs/advanced/custom-components.mdx:
````diff
@@ -0,0 +1,141 @@
+    )
+}
+```
+Since this is also a regular typescript class, you can add methods. This is especially useful
````
Did you decide not to use "also" here?

nick-1666 replied:
Since this is still... was transformed into Since this is also. is that not what you meant?

hhenrichsen replied:
I was intending "Since this is a regular typescript class, you can also add methods."

nick-1666 replied:
should typescript be omitted since its implied that typescript is used throughout the project?

hhenrichsen replied:
I think typescript is a good distinction because we're talking about component classes as well, so having a distinction between a "regular TypeScript class" and the rules for a component class is good. We might capitalize TypeScript, though.

hhenrichsen reviewed (commented):
Again some more grammar things I found, but not going to block on any of them.

nick-1666 commented:
fixed-up

hhenrichsen reviewed (commented):
Couple extra thoughts but no blockers.

nick-1666 commented on packages/docs/docs/advanced/custom-components.mdx:
```diff
@@ -0,0 +1,140 @@
+Notice how colors are wrapped in `ColorSignal<>` whilst any other type (even user-defined ones)
+are wrapped in `SimpleSignal<>`. The type does not need to be passed to color signal as Motion
+Canvas knows that it must be of a color-resolvable type. In both, the class is passed at the end
+of the wrapper to register the signal to the class. The properties must also be _declared_, _public_ 
```
no cause its not an actual key word

nick-1666 replied:
i dont think so. If we wanted to use keywords the sentence could be changed to:
"The properties must be (initialised/initialized depending on uk or us) with the `public` `declare` and `readonly` keywords"

nick-1666 replied:
or better yet just "Properties ..."

nick-1666 commented:
this should have fixed the formatting
